jobs/build/ocp4/build.groovy

```diff
@@ -0,0 +1,606 @@
+                list.add(it["name"])
+            }
+        }
+    }
+    return changed
+}
+
+// determine what doozer parameter (if any) to use for includes/excludes
+def includeExclude(kind, includes, excludes) {
+    // --latest-parent-version only applies for images but won't hurt for RPMs
+    if (includes) { return "--latest-parent-version --${kind} ${includes}" }
+    if (excludes) { return "--latest-parent-version --${kind} '' --exclude ${excludes}" }
+    return "--${kind} ''"
+}
+
+def stageBuildRpms() {
+    if (!buildPlan.buildRpms) {
+        echo "Not building RPMs."
+        return
+    }
+    def cmd =
+        """
+        ${doozerOpts}
+        ${includeExclude "rpms", buildPlan.rpmsIncluded, buildPlan.rpmsExcluded}
+        rpms:build --version v${version.full}
+        --release ${version.release}
+        """
+
+    buildPlan.dryRun ? echo("doozer ${cmd}") : buildlib.doozer(cmd)
+}
+
+/**
+ * If necessary, puddle a new compose from our candidate tag.
+ * Set the new puddle directory (e.g. "YYYY-MM-DD.1")
+ */
+def stageBuildCompose() {
+
+    // we may or may not have (successfully) built the openshift RPM in this run.
+    // in order to script the correct version to publish later, determine what's there now.
+    finalRpmVersionRelease = latestOpenshiftRpmBuild(version.branch).replace("openshift-", "")
+
+    if (!buildPlan.buildRpms && !buildPlan.forceBuild) {
+        // a force build of just images is likely to want to pick up new dependencies,
+        // so in that case still create the compose.
+        echo "No RPMs built, not a force build; no need to create a compose"
+        return
+    }
+    if(buildPlan.dryRun) {
+        echo "Build puddle with conf ${buildPlan.puddleConf}"
+        rpmMirror.composeDir = "YYYY-MM-DD.1"
+        return
+    }
+
+    rpmMirror.composeDir = buildlib.build_puddle(
+        buildPlan.puddleConf,    // The puddle configuration file to use
+        null,   // signing key
+        "-b",   // do not fail if we are missing dependencies
+        "-d",   // print debug information
+        "-n",   // do not send an email for this puddle
+        "-s",   // do not create a "latest" link since this puddle is for building images
+        "--label=building"   // create a symlink named "building" for the puddle
+    )
+}
+
+def stageUpdateDistgit() {
+    if (!buildPlan.buildImages) {
+        echo "Not rebasing images."
+        return
+    }
+    def cmd =
+        """
+        ${doozerOpts}
+        ${includeExclude "images", buildPlan.imagesIncluded, buildPlan.imagesExcluded}
+        images:rebase --version v${version.full} --release ${version.release}
+        --message 'Updating Dockerfile version and release v${version.full}-${version.release}' --push
+        """
+    if(buildPlan.dryRun) {
+        echo "doozer ${cmd}"
+        return
+    }
+    buildlib.doozer(cmd)
+    // TODO: if rebase fails for required images, notify image owners, and still notify on other reconciliations
```

Comment:
Don't want to forget about these 2 TODOs. That's referring to how we are notifying image owners presently right?

Reply:
Right. It requires a little doozer work that I haven't prioritized yet, because right now if `doozer images:rebase` fails for a required image, then we stop the build, and so notifications don't go out for what was successfully rebased.

I put it off because I don't think reconciliation notices are that critical.